File: src/riscv/rvprivileged.py
# Copyright 2026 Flavien Solt, Quentin Bordier ETH Zurich.
# Licensed under the General Public License, Version 3.0, see LICENSE for details.
# SPDX-License-Identifier: GPL-3.0-only


from .rvprotoinstrs import *

RV32I_OPCODE_PRIVILEGED = 0b1110011

# All functions return uint32_t


def rvprivileged_sret():
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0b00010, 0b0001000)


def rvprivileged_mret():
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0b00010, 0b0011000)


def rvprivileged_wfi():
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0b00101, 0b0001000)


def rvprivileged_sfence_vma(rs1: int, rs2: int):
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, rs1, rs2, 0b0001001)


def rvprivileged_sinval_vma(rs1: int, rs2: int):
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, rs1, rs2, 0b0001011)


def rvprivileged_sfence_w_inval():
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0b00000, 0b0001100)


def rvprivileged_sfence_inval_ir():
    return instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0b00001, 0b0001100)

File: go/riscv/rvprivileged.go
// Copyright 2026 Flavien Solt, Quentin Bordier ETH Zurich.
// SPDX-License-Identifier: GPL-3.0-only

package riscv

// Bit encoders preserve the original argument order and immediate masking.
const RV32I_OPCODE_PRIVILEGED int32 = 115

func Rvprivileged_sret() uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 2, 8)
}

func Rvprivileged_mret() uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 2, 24)
}

func Rvprivileged_wfi() uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 5, 8)
}

func Rvprivileged_sfence_vma(rs1 int32, rs2 int32) uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, rs1, rs2, 9)
}

func Rvprivileged_sinval_vma(rs1 int32, rs2 int32) uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, rs1, rs2, 11)
}

func Rvprivileged_sfence_w_inval() uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 0, 12)
}

func Rvprivileged_sfence_inval_ir() uint32 {
	return Instruc_rtype(RV32I_OPCODE_PRIVILEGED, 0, 0, 0, 1, 12)
}
